nemo/collections/nlp/data/datasets/text_classification_dataset.py

```diff
@@ -52,68 +61,102 @@ class BertTextClassificationDataset(Dataset):
 
-                all_sent_subtokens.append(sent_subtokens)
-                sent_lengths.append(len(sent_subtokens))
+                if shuffle or num_samples > -1:
+                    random.seed(0)
+                    random.shuffle(lines)
+                    if num_samples > 0:
+                        lines = lines[:num_samples]
 
-        get_stats(sent_lengths)
-        self.max_seq_length = min(max_seq_length, max(sent_lengths))
+                for index, line in enumerate(lines):
+                    if index % 20000 == 0:
+                        logging.debug(f"Processing line {index}/{len(lines)}")
 
-        for i in range(len(all_sent_subtokens)):
-            if len(all_sent_subtokens[i]) > self.max_seq_length:
-                shorten_sent = all_sent_subtokens[i][-self.max_seq_length + 1 :]
-                all_sent_subtokens[i] = ['[CLS]'] + shorten_sent
-                too_long_count += 1
+                    sent_label = int(line.split()[-1])
+                    sent_labels.append(sent_label)
+                    sent_words = line.strip().split()[:-1]
+                    sent_subtokens = ['[CLS]']
 
-        logging.info(
-            f'{too_long_count} out of {len(sent_lengths)} \
-                       sentencess with more than {max_seq_length} subtokens.'
-        )
+                    for word in sent_words:
+                        word_tokens = tokenizer.tokenize(word)
+                        sent_subtokens.extend(word_tokens)
 
-        self.convert_sequences_to_features(all_sent_subtokens, sent_labels, tokenizer, self.max_seq_length)
+                    sent_subtokens.append('[SEP]')
 
-        self.tokenizer = tokenizer
-        self.vocab_size = self.tokenizer.vocab_size
+                    all_sent_subtokens.append(sent_subtokens)
+                    sent_lengths.append(len(sent_subtokens))
+
+            get_stats(sent_lengths)
+
+            for i in range(len(all_sent_subtokens)):
+                if len(all_sent_subtokens[i]) > max_seq_length:
+                    shorten_sent = all_sent_subtokens[i][-max_seq_length + 1 :]
+                    all_sent_subtokens[i] = ['[CLS]'] + shorten_sent
+                    too_long_count += 1
+
+            logging.info(
+                f'{too_long_count} out of {len(sent_lengths)} \
+                        sentences with more than {max_seq_length} subtokens.'
+            )
+
+            self.convert_sequences_to_features(all_sent_subtokens, sent_labels, tokenizer, max_seq_length)
+
+            if self.use_cache:
+                self.cache_features(hdf5_path, self.features)
+
+                # update self.features to use features from hdf5
+                self.load_cached_features(hdf5_path)
 
     def __len__(self):
-        return len(self.features)
+        if self.use_cache:
+            return len(self.features[0])
+
+        else:
+            return len(self.features)
 
     def __getitem__(self, idx):
+        if self.use_cache:
```

Comment:
@VahidooX what do you think about this set up?

Reply:
I think it should be ok for now. We can merge it into nemo and improve it later.